workflow/lib/preprocess/preprocess.py

```diff
@@ -171,3 +605,479 @@ def nd2_to_tiff(
+    """
+    if data_format not in DATA_FORMATS:
+        raise ValueError(f"Unsupported data format: {data_format}")
+    if data_organization not in DATA_ORGANIZATIONS:
+        raise ValueError(f"Unsupported data organization: {data_organization}")
+
+    if isinstance(file_paths, str):
+        file_paths = [file_paths]
+
+    if data_format == "nd2":
+        if data_organization == "tile":
+            # For tile organization, process each file separately
+            metadata_dfs = []
+            for i, file_path in enumerate(file_paths):
+                current_tile = tile if tile is not None else i
+
+                # Only pass parameters that are not None
+                kwargs = {
+                    "file_path": file_path,
+                    "plate": plate,
+                    "well": well,
+                    "tile": current_tile,
+                    "verbose": verbose,
+                }
+
+                if cycle is not None:
+                    kwargs["cycle"] = cycle
+                if round is not None:
+                    kwargs["round"] = round
+
+                df = extract_metadata_tile_nd2(**kwargs)
+                metadata_dfs.append(df)
+            return pd.concat(metadata_dfs, ignore_index=True)
+
+        elif data_organization == "well":
+            # For well organization, process the first file (should be only one)
+            kwargs = {
+                "file_path": file_paths[0],
+                "plate": plate,
+                "well": well,
+                "verbose": verbose,
+            }
+
+            if cycle is not None:
+                kwargs["cycle"] = cycle
+            if round is not None:
+                kwargs["round"] = round
+
+            return extract_metadata_well_nd2(**kwargs)
+
+    elif data_format == "tiff":
+        # For TIFF, always treat as tile-based for now
+        metadata_dfs = []
+        for i, file_path in enumerate(file_paths):
+            current_tile = tile if tile is not None else i
+
+            # Only pass parameters that are not None
+            kwargs = {
+                "file_path": file_path,
+                "plate": plate,
+                "well": well,
+                "tile": current_tile,
+                "metadata_file_path": metadata_file_path,
+                "verbose": verbose,
+            }
+
+            if cycle is not None:
+                kwargs["cycle"] = cycle
+            if round is not None:
+                kwargs["round"] = round
+
+            df = extract_metadata_tiff(**kwargs)
+            metadata_dfs.append(df)
+        return pd.concat(metadata_dfs, ignore_index=True)
+
+
+def convert_to_array(
+    files: Union[str, List[str], Path, List[Path]],
+    data_format: str = "nd2",
+    data_organization: str = "tile",
+    position: int = None,
+    channel_order_flip: bool = False,
+    verbose: bool = False,
+    **kwargs,
+) -> np.ndarray:
+    """Convert microscopy image files to numpy array in CYX format.
```

Comment:
This is awesome :D

Reply:
I really like handling all of the file differences and such upstream so we can work with consistent data downstream. Let's try to keep doing this 🔥